### PR TITLE
Show full flow error text in linter description

### PR DESCRIPTION
**Pull request by lloiser**
image

Requires: https://github.com/steelbrain/linter-ui-default/pull/342

### CODE REVIEW

File: lib/helpers-linter.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+// https://github.com/facebook/flow/blob/v0.47.0/tsrc/flowResult.js
+// and adjusted to output nicely formatted markdown.
+
+import { format } from 'util'
+import type { StatusError, StatusExtra, StatusMessage, Location } from './types'
+
+function fileUrl({ source, start }: Location): string {
+  // build a link that can be opened by the linter ui.
+  // e.g. atom://linter?file=<path>&row=<number>&column=<number>
+  return 'atom://linter?file=' + source +
```
Comment (steelbrain) [inline lib/helpers-linter.js]: Can we url encode the source please?

File: lib/helpers-linter.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    (start.line > 0
```
Comment (steelbrain) [inline lib/helpers-linter.js]: Instead of doing it inline can we do it in a variable above this line for verbosity?

**Comment by nodkz** [comment]
@lloiser @steelbrain hi guys

Can we improve with this PR displaying errors in several locations (different files)?

I use Flow for React components, and when I miss required property for component (in `SwipeMagicRenderer.js`)

screen shot 2017-06-21 at 20 10 29

the error displays only in a file `SwipeableResizer.js` where the component is defined:

screen shot 2017-06-21 at 20 11 01

Flowtype provides the following information for this Error ([full json output gist](https://gist.github.com/nodkz/c6312920c1284f984a91010f3a037740)):
```js
{
  message: [
    { loc: { source: '..../SwipeableResizer.js' } },
    { loc: undefined },
    { loc: { source: '..../SwipeMagicRenderer.js' } },
  ]
}
```

But in function [toLinterReference](https://github.com/steelbrain/flow-ide/pull/94/files#diff-27a1c86133de4310373decfdad200a91R27) it takes only first file. But for better developer experience will be better to display error in all different location:
```js
function toLinterReference(messages) {
  for (let i = 1, length = messages.length; i < length; i++) {
    const message = messages[i]
    if (message.loc) {
      return {
        file: message.loc.source,
        position: locationToRange(message.loc)[0],
      }
    }
  }
  return null
}
```

PS. Similar issue for Nuclide I [opened 1 year ago](https://github.com/facebook/nuclide/issues/574#issuecomment-232289879), but they did not understand me. 

PSS `flow-ide` plugin is much-much better. And can be better even more. Thanks for such awesome thing!

**Comment by steelbrain** [comment]
Hi @nodkz 

I had implemented what you're saying in my Atom Hack package, which's linting part was seperated and is now the linter package, I dropped it for plenty of reasons.

The first and most important one is how do you distinguish between a normal message and a part of any other message? Keep in mind that blue underlines are already part of Info messages, also what do we show there? All of the message or just the relevant part? If all, how's the user going to know which part is the current one.

The second most important reason is performance, every new added message will be searched for references for not just the current text editor but messages from all text editors and will be added/removed when the text editor is closed and opened up again so this scales pretty horribly.

Also the support for standardized traces was removed in Linter Messages v2 to allow for a markdown string so v2 providers are inherently incapable of such a thing

**Comment by lloiser** [comment]
@nodkz I would agree with you _if_ flow would provide a better error message for each entry in `message`. Unfortunately flow's errors often only make sense if you read the complete error which is now visible in the error description.
I will merge this now. If you still think your suggestion is valid please open another issue and/or a pull request.

**Comment by nodkz** [comment]
@lloiser working with your full error toggling about one week. This is amazing thing!!! Thanks a lot. I almost forget about running flow via CLI.

And of course, @steelbrain thank you too, for all your outstanding plugins.

**Pull request merged**